BSB_LAN/BSB_LAN.ino

```diff
@@ -7349,22 +7349,27 @@ uint8_t pps_offset = 0;
           }
           // if the file is available, read from it:
           if (dataFile) {
+            unsigned long filesize = dataFile.size();
+#if !defined(ESP32)
             dir_t d;
             uint16_t lastWrtYr = 0;
             byte monthval = 0;
             byte dayval = 0;
-            unsigned long filesize = dataFile.size();
             if (dataFile.dirEntry(&d)) {
               lastWrtYr = FAT_YEAR(d.lastWriteDate);
               monthval = FAT_MONTH(d.lastWriteDate);
               dayval = FAT_DAY(d.lastWriteDate);
               }
-
+#endif
             if ((httpflags & HTTP_ETAG))  { //Compare ETag if presented
+#if !defined(ESP32)
               if (memcmp(outBuf, outBuf + buffershift, sprintf_P(outBuf + buffershift, PSTR("\"%02d%02d%d%02d%02d%02d%lu\""), dayval, monthval, lastWrtYr, FAT_HOUR(d.lastWriteTime), FAT_MINUTE(d.lastWriteTime), FAT_SECOND(d.lastWriteTime), filesize))) {
                 // reuse httpflags
                 httpflags &= ~HTTP_ETAG; //ETag not match
               }
+#else
+                httpflags &= ~HTTP_ETAG; //ETag not match
```

Comment:
Indention (or code?) is wrong

Reply:
Can you tell me a bit more about the errors and how they are related to the SD library? Because I don't get these errors here (or do you mean warnings?).
Also, I'm using Ethernet on the Olimex and the speed is really fine.
Serial timeout errors come whenever there is a character received on the serial port and more than a second goes by without a subsequent character (until a complete command has been received). Is there something sending on your serial port when you use the terminal program? Maybe some kind of keep-alive keypress sent from the terminal program?

Reply:
>Can you tell me a bit more about the errors and how they are related to the SD library? Because I don't get these errors here (or do you mean warnings?).

I call some functions and use macroses from sdfat library for file date and time request for WEBSERVER module. These functions are specific for sdfat and not exist in SD_MMC library for ESP32. Also i tried run sdfat on EVB board but unsuccessful.

>Also, I'm using Ethernet on the Olimex and the speed is really fine.

:-( Sometimes perfomance is great but time-after-time MCU "hangs": ICMP packets are lost and "Serial input timeout" appear.

>Serial timeout errors come whenever there is a character received on the serial port and more than a second goes by without a subsequent character (until a complete command has been received). Is there something sending on your serial port when you use the terminal program? Maybe some kind of keep-alive keypress sent from the terminal program?

No, i tried to connect USB cable to power supply directly but "Serial input timeout" are continued to appear.
Now I think about SD card replacing.

Reply:
Ah, ok, got it - since I'm not using the WEBSERVER module, I didn't encounter this error.
Yes, SdFat does not work on the ESP, so I had to find an alternative which was SD_MMC. 

As for your errors, it's really strange regarding the serial stuff. As for the Ethernet performance, there is a collision between the use of the SD card and Ethernet which is supposed to be fixed in the most recent ESP32 framework.
These lines in SD_MMC.cpp are worth a check:
```
    host.max_freq_khz = SDMMC_FREQ_HIGHSPEED;
#ifdef BOARD_HAS_1BIT_SDMMC
    mode1bit = true;
#endif
    if(mode1bit) {
        host.flags = SDMMC_HOST_FLAG_1BIT; //use 1-line SD mode
	slot_config.width = 1;
    }
```
On the Olimex, you have to use 1-line SD mode if you also use Ethernet...

Reply:
"Serial input timeout" was caused by OneWire library on GPIO 3 (yes, i stupid sometime).
WiFi works fine, but ethernet still unstable. BOARD_HAS_1BIT_SDMMC is defined in .h files so here no problem.
Next stages: a) wil try to flash simple ethernet sketch to board and check connection. b) will try to connect dallas sensor to GPIO 13 or 16 (they looks unused right now).

Reply:
Yes, it's a bit confusing that the pin numbering has been changed in such a way, I had my problems there, too... The ones you mentioned are free only if you don't use any I2C devices. The EVB has almost every pin use/connected in some way ;)...

As for the SD card problems: Have you tried SD card with WiFi and no SD card (also disabled in BSB_LAN) with Ethernet? If both of these run fine, then there should still be some kind of issue with the SD card configuration. I _think_ I also made some other changes in SD_MMC.cpp where I reduced transmission speed or something like that.

Reply:
> The ones you mentioned are free only if you don't use any I2C devices.

Yep.  Board with lot of contacts misled me from start. :-)

> Have you tried SD card with WiFi 

This configuration works excellent! 

> no SD card (also disabled in BSB_LAN) with Ethernet?

I tried it. Unfortunately Ethernet still glitchy.

By the way i tried today esp 2.0.0 SDK but it config looks not suitable for us now: some macroses and data types not defined, ex. rtc_gpio_desc; For Linux users here is no problem (make menuconfig and so on), but Windows users with Arduino Studio will get headashe.

Reply:
> If both of these run fine, then there should still be some kind of issue with the SD card configuration. I think I also made some other changes in SD_MMC.cpp where I reduced transmission speed or something like that.

Can you do diff?

Reply:
Can't do a diff anymore, but here's the SD_MMC.cpp I'm using
[SD_MMC.cpp.zip](https://github.com/fredlcore/BSB-LAN/files/7180095/SD_MMC.cpp.zip)

Reply:
Thank you!

Reply:
Maybe you can run a diff and see if there are any differences?

Reply:
File from distrib and your file are identical.
But. I tried to flash program without these strings
```
//    #define ETH_CLK_MODE ETH_CLOCK_GPIO17_OUT
//    #define ETH_PHY_POWER 12 
```
and ethernet run very nice!

Reply:
I found solution: https://github.com/espressif/arduino-esp32/issues/495#issuecomment-314980250

i replaced  host.max_freq_khz = SDMMC_FREQ_HIGHSPEED;
with     host.max_freq_khz = SDMMC_FREQ_PROBING;

Reply:
Yes, I think I changed this line, too, at some point, but at least here it runs fine without any adjustments as long as the SDMMC_HOST_FLAG_1BIT flag is set.
Have you figured out what the difference is between SDMMC_FREQ_HIGHSPEED and SDMMC_FREQ_PROBING? And could host.flags be set from within BSB-LAN without modifying the library code?

Reply:
Perhaps we have different board revisions. LAN began work stable when i set SDMMC_FREQ_PROBING: no lost pings or delays. I did not performance measures but summary (SD reading + LAN throughput) it looks much faster than Due. Unfortunately i found not any ways to change this setting from sketch. I can propose to make a patch for SDK and tried to do pull request or open issue on espressif/arduino-esp32. As alternative way we can add this case in manual and add patched SDMMC library to tree.

Frequency settings at bottom of page: https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/storage/sdmmc.html

P.S. ETH_CLK_MODE and ETH_PHY_POWER  still commented.

Reply:
I have revision "I", I think a PR/issue for espressif/arduino-esp32 would be great, maybe you could try setting it to SDMMC_FREQ_52M and SDMMC_FREQ_26M as well to figure out how slow the speed needs to be set.
And could you also try and set ETH_CLK_MODE and ETH_PHY_POWER back to normal to check if these changes are necessary as well? I assume that the reduced clock speed should solve the problem, but better be certain...

Reply:
Also, have you seen this?
https://github.com/espressif/arduino-esp32/issues/495#issuecomment-314980250

Reply:
Also, have you tried different SD cards? On the Ethernet Shield for the Arduino, we've had this as an issue before...

Reply:
Results:
Attempt 1. 
SDMMC_FREQ_PROBING

Result:
Stable.

Attempt 2.  
ETH_CLK_MODE ETH_CLOCK_GPIO17_OUT
ETH_PHY_POWER 12 
SDMMC_FREQ_PROBING

Result:
Unstable LAN + unstable SD
10:09:27.348 -> E (126232) sdmmc_cmd: sdmmc_read_sectors_dma: sdmmc_send_cmd returned 0x107
10:09:27.348 -> E (126233) diskio_sdmmc: sdmmc_read_blocks failed (263)

Attempt 3.  
ETH_CLK_MODE ETH_CLOCK_GPIO17_OUT
ETH_PHY_POWER 12 
SDMMC_FREQ_52M

Result:
Unstable LAN + unstable SD
10:15:47.812 -> E (246232) sdmmc_cmd: sdmmc_read_sectors_dma: sdmmc_send_cmd returned 0x107
10:15:47.812 -> E (246233) diskio_sdmmc: sdmmc_read_blocks failed (263)

Attempt 4.  
ETH_CLK_MODE ETH_CLOCK_GPIO17_OUT
ETH_PHY_POWER 12 
SDMMC_FREQ_26M

Result:
Fail.
0:19:31.488 -> Starting SD..assertion "card->max_freq_khz <= card->host.max_freq_khz" failed: file "/home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/sdmmc/sdmmc_common.c", line 187, function: sdmmc_init_host_frequency
10:19:31.568 -> abort() was called at PC 0x40103d13 on core 1
10:19:31.568 -> 
10:19:31.568 -> ELF file SHA256: 0000000000000000
10:19:31.568 -> 
10:19:31.568 -> Backtrace: 0x40089eec:0x3ffb1d40 0x4008a169:0x3ffb1d60 0x40103d13:0x3ffb1d80 0x4016ab42:0x3ffb1db0 0x4016a420:0x3ffb1de0 0x40132c56:0x3ffb1e00 0x400e64c9:0x3ffb1e50 0x400d84fe:0x3ffb1ed0 0x400e8d32:0x3ffb1fb0 0x4008b17a:0x3ffb1fd0
10:19:31.568 -> 
10:19:31.568 -> Rebooting...

Attempt 5.  
SDMMC_FREQ_26M

Result:
Fail.
10:23:03.603 -> Starting SD..assertion "card->max_freq_khz <= card->host.max_freq_khz" failed: file "/home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/sdmmc/sdmmc_common.c", line 187, function: sdmmc_init_host_frequency
10:23:03.683 -> abort() was called at PC 0x40103cfb on core 1
10:23:03.683 -> 
10:23:03.683 -> ELF file SHA256: 0000000000000000
10:23:03.683 -> 
10:23:03.683 -> Backtrace: 0x40089eec:0x3ffb1d40 0x4008a169:0x3ffb1d60 0x40103cfb:0x3ffb1d80 0x4016ab2a:0x3ffb1db0 0x4016a408:0x3ffb1de0 0x40132c3e:0x3ffb1e00 0x400e64b1:0x3ffb1e50 0x400d8511:0x3ffb1ed0 0x400e8d1a:0x3ffb1fb0 0x4008b17a:0x3ffb1fd0
10:23:03.723 -> 
10:23:03.723 -> Rebooting...

Attempt 6.  
SDMMC_FREQ_52M

Result:
Stable LAN, SD failed after some readings.
10:26:14.429 -> Setup complete
10:26:28.830 -> E (59433) sdmmc_cmd: sdmmc_read_sectors_dma: sdmmc_send_cmd returned 0x107
10:26:28.830 -> E (59433) diskio_sdmmc: sdmmc_read_blocks failed (263)

Reply:
Ok, I have found this thread which mentions 
I'm a bit puzzled what this message means:
```
card->max_freq_khz <= card->host.max_freq_khz
```
Why does it complain when you set the frequency to 26M, but not when it's set to 52M?
SDMMC_FREQ_PROBING is just 400(k?)Hz, so it's way slower than the other modes.

And it seems to me that disabling the defines still make it work on my machine here, so maybe the code has been updated so that it is no longer necessary/advisable to have these defines in the code.#

So it seems that it is just the host.max_freq_khz that is relevant here. If you could open an issue/PR with the espressif people, we could hope for a fixed or more user-defined approach. Otherwise, we'll have to provide our own library (which would suck a bit)...

Reply:
> Why does it complain when you set the frequency to 26M, but not when it's set to 52M?

May this values hard linking with quartz.

> If you could open an issue/PR with the espressif people, we could hope for a fixed or more user-defined approach. 

I will try to do it.

Reply:
Tried to run with SDMMC_FREQ_DEFAULT - works fine, much faster (at least 10 times) than SDMMC_FREQ_PROBING.

Reply:
Huh? Why does it work now with SDMMC_FREQ_DEFAULT? This is the standard setting, so does it mean that no change is necessary?
Also, did you try a different SD card? Maybe that's part of the problem?

Reply:
> Why does it work now with SDMMC_FREQ_DEFAULT? This is the standard setting, so does it mean that no change is necessary?

No. in v. 1.0.6 and 2.0.0 value host.max_freq_khz set twice: first time as host.max_freq_khz = SDMMC_FREQ_DEFAULT; and second time host.max_freq_khz = SDMMC_FREQ_HIGHSPEED; . Without conditions, switches, definitions, so effectife value for host.max_freq_khz is always SDMMC_FREQ_HIGHSPEED. 
I searched in the history and saw that this value was originally like this.

Reply:
You are right, this really does not make sense. Maybe they could at least use a ifdef or something like that, because the default should be the default ;). 
But maybe it's still worth trying a different SD card to see if the problem is Olimex-related or SD-card related...